pkg/controllers/management/feature/feature_handler.go

```diff
@@ -0,0 +1,104 @@
+package feature
+
+import (
+	"context"
+	"time"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/rancher/pkg/features"
+	managementv3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/wrangler"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+type handler struct {
+	featuresClient managementv3.FeatureClient
+	tokensLister   managementv3.TokenCache
+	tokenEnqueue   func(string, time.Duration)
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) {
+	h := handler{
+		featuresClient: wContext.Mgmt.Feature(),
+		tokensLister:   wContext.Mgmt.Token().Cache(),
+		tokenEnqueue:   wContext.Mgmt.Token().EnqueueAfter,
+	}
+	wContext.Mgmt.Feature().OnChange(ctx, "feature-handler", h.sync)
+}
+
+func (h *handler) sync(_ string, obj *v3.Feature) (*v3.Feature, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	obj, err := h.setLockedValue(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	if obj.Name == features.TokenHashing.Name() {
```

Comment:
It looks below like you call `refreshTokens` even if it wasn't updated ? 
```
	if !update {		return obj, nil	}
```
error is nil so this still gets hit, why refresh ?

Reply:
Because the update check is for for LockedValue where you want to refresh tokens if the effective value of the feature flag is true. This could be the regular Value field or LockedValue.